### PR TITLE
Add defaults openapi

### PR DESCRIPTION
**Pull request by apelisse**
**What type of PR is this?**
Generate built-in defaults field in OpenAPI based on the marker. Technically doesn't really changes anything since we're still publishing the same thing, and we're not using the marker yet to generate the defaulting code. But at least server-side apply will have access to the defaults now.

/kind feature

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by apelisse** [comment]
I suspect this is failing because of https://github.com/kubernetes/test-infra/issues/19910

File: staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/controller.go [hunk 1]
```diff
@@ -202,7 +202,7 @@ func buildVersionSpecs(crd *apiextensionsv1.CustomResourceDefinition, oldSpecs m
 		if !v.Served {
 			continue
 		}
-		spec, err := builder.BuildSwagger(crd, v.Name, builder.Options{V2: true, StripDefaults: true})
+		spec, err := builder.BuildSwagger(crd, v.Name, builder.Options{V2: true, StripDefaults: false})
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/controller.go:205]: where are these defaults stripped? This controller is the one creating the spec for later publishing.

Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/controller.go:205]: Is it https://github.com/kubernetes/kubernetes/pull/96480/commits/fb7a94bc740997452b8eb64fc765ff8b4decba40#diff-53cefa59736af8bf2b6669d4041bc7c6dacc08400f32e2348fcf42d0b0b180a5R193 in the aggregator that takes over the pruning? Maybe add a comment here.

File: staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator/internal/default_pruning.go [hunk 2]
```diff
@@ -0,0 +1,206 @@
+/*
```
Comment (sttts) [inline staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator/internal/default_pruning.go:1]: is this a copy of the algorithm used by builder? Can we merge them?

Comment (apelisse) [reply staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator/internal/default_pruning.go:1]: not really, the builder goes from CRD type representation to `swagger.Spec`. This is operating directly on the swagger`.Spec`. On the other hand, I don't think there is any call left to the builder with StripDefaults to true, so we might be able to remove it.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/controller.go (hunk 1 above)
Comment (apelisse) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/controller.go:205]: Yeah, we're pruning before serving, and keeping all the "internal representation" with the defaults.

Comment (apelisse) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/controller.go:206]: I've added a comment.

File: staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator/internal/default_pruning.go (hunk 2 above)
Comment (apelisse) [reply staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator/internal/default_pruning.go:1]: OK I just removed it!

**Comment by Jefftree** [comment]
/lgtm
The new change just fixes an issue with the bazel build.

File: pkg/generated/openapi/openapi_test.go [hunk 3]
```diff
@@ -24,12 +24,15 @@ import (
...
 	"k8s.io/apimachinery/pkg/util/diff"
+	"k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator"
 )
 
 func TestOpenAPIRoundtrip(t *testing.T) {
 	dummyRef := func(name string) spec.Ref { return spec.MustCreateRef("#/definitions/dummy") }
 	for name, value := range GetOpenAPIDefinitions(dummyRef) {
 		t.Run(name, func(t *testing.T) {
+			// TODO(kubernetes/gengo#193): Remove this line
+			value.Schema = *aggregator.PruneDefaultsSchema(&value.Schema)
```
Comment (sttts) [inline pkg/generated/openapi/openapi_test.go:35]: add the reason for removal here please.

File: staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator/aggregator.go [hunk 4]
```diff
@@ -189,6 +189,7 @@ func (s *specAggregator) buildOpenAPISpec() (specToReturn *spec.Swagger, err err
 			specToReturn.Paths = nil
 			specToReturn.Definitions = nil
 		}
+		specInfo.spec.Definitions = PruneDefaults(specInfo.spec.Definitions)
```
Comment (sttts) [inline staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator/aggregator.go:192]: Aggregated API servers won't call this. Hence, they will expose the defaults. But as this goes through the aggregrator before the clients see it, it's probably ok.

**Comment by sttts** [comment]
Nits.

LGTM otherwise.

File: staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator/aggregator.go (hunk 4 above)
Comment (apelisse) [reply staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator/aggregator.go:192]: Thanks, I think that's the reason for the current test failure :-)

File: staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator/aggregator.go [hunk 5]
```diff
@@ -189,6 +189,7 @@ func (s *specAggregator) buildOpenAPISpec() (specToReturn *spec.Swagger, err err
 			specToReturn.Paths = nil
 			specToReturn.Definitions = nil
 		}
+		specInfo.spec.Definitions = handler.PruneDefaults(specInfo.spec.Definitions)
```
Comment (liggitt) [inline staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator/aggregator.go:192]: does this mutate data held by `s.openAPISpecs`? if so, is specAggregator free to mutate that?

Comment (apelisse) [reply staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator/aggregator.go]: Thanks, I ended-up doing a copy, and pruning a little earlier.

**Comment by liggitt** [comment]
/lgtm

/milestone v1.20
per https://groups.google.com/g/kubernetes-sig-release/c/eCk4Xyb58GQ/m/LM9Xsy0hAgAJ

**Comment by liggitt** [comment]
flake #96565 
/retest

**Comment by apelisse** [comment]
Rebased!

**Comment by Jefftree** [comment]
/retest

```
wrapper.sh] [TEST] Running Test Command: `bash -c curl -sSL https://kind.sigs.k8s.io/dl/latest/linux-amd64.tgz | tar xvfz - -C "${PATH%%:*}/" && e2e-k8s.sh` ...
curl: (56) OpenSSL SSL_read: SSL_ERROR_SYSCALL, errno 104
```

**Pull request merged**